File: sdks/go/tracing.go
package amlexia

import (
	"crypto/rand"
	"encoding/hex"
)

type TraceContext struct {
	TraceID        string
	SpanID         string
	ParentSpanID   string
	SessionID      string
	UserID         string
	Environment    string
	ReleaseVersion string
}

func randomHex(n int) string {
	b := make([]byte, n)
	_, _ = rand.Read(b)
	return hex.EncodeToString(b)
}

func CreateTraceContext() TraceContext {
	return TraceContext{
		TraceID: randomHex(16),
		SpanID:  randomHex(8),
	}
}

func ChildSpan(parent TraceContext) TraceContext {
	return TraceContext{
		TraceID:        parent.TraceID,
		SpanID:         randomHex(8),
		ParentSpanID:   parent.SpanID,
		SessionID:      parent.SessionID,
		UserID:         parent.UserID,
		Environment:    parent.Environment,
		ReleaseVersion: parent.ReleaseVersion,
	}
}

File: sdks/node/src/tracing.ts
import { newSpanId, newTraceId } from '@amlexiahq/shared';
import type { TrackEvent } from './types.js';

export interface TraceContext {
  traceId: string;
  spanId: string;
  parentSpanId?: string;
  sessionId?: string;
  userId?: string;
  environment?: string;
  releaseVersion?: string;
}

export function createTraceContext(options?: Partial<TraceContext>): TraceContext {
  return {
    traceId: options?.traceId ?? newTraceId(),
    spanId: options?.spanId ?? newSpanId(),
    parentSpanId: options?.parentSpanId,
    sessionId: options?.sessionId,
    userId: options?.userId,
    environment: options?.environment ?? process.env.NODE_ENV,
    releaseVersion: options?.releaseVersion ?? process.env.AMLEXIA_RELEASE,
  };
}

export function childSpan(parent: TraceContext): TraceContext {
  return {
    ...parent,
    spanId: newSpanId(),
    parentSpanId: parent.spanId,
  };
}

export function applyTraceToEvent(event: TrackEvent, ctx: TraceContext): TrackEvent {
  return {
    ...event,
    traceId: ctx.traceId,
    spanId: ctx.spanId,
    parentSpanId: ctx.parentSpanId,
    sessionId: ctx.sessionId ?? event.sessionId,
    userId: ctx.userId ?? event.userId,
    environment: ctx.environment ?? event.environment,
    releaseVersion: ctx.releaseVersion ?? event.releaseVersion,
  };
}
